### PR TITLE
[STACK-2399] Fix set Loadbalancer error

### PR DESCRIPTION
**Pull request by ytsai-a10**
If Bug Fix:
Severity Level: Critical
Issue Description: openstack loadbalancer set will failed, since loadbalancer_update flow missing parameters in store for HandleVRIDFloatingIP task.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2399

## Technical Approach
- Provide proper a10constants.VTHUNDER_CONFIG and a10constants.USE_DEVICE_FLAVOR for vthunder update flow.
- create vrid if it is not exist on thunder.

## Config Changes
<pre>
<b>[a10_global]
network_type = vlan
vrid=7
vrid_floating_ip = ".126"
#use_parent_partition=True

[a10_controller_worker]
network_driver = a10_octavia_neutron_driver
loadbalancer_topology = SINGLE

[listener]
autosnat = True
conn_limit=20000

[health_monitor]
post_data = "abc=1"

[a10_house_keeping]
#use_periodic_write_memory = 'enable'
write_mem_interval = 300

[hardware_thunder]
devices = [
                    {
                     "project_id": "3d21c71c4e4040599933bda62a76ae2f",
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     "device_name": "admin"
                     },
                    {
                     "ip_address": "192.168.90.46",
                     "username": "admin",
                     "password": "a10",
                     #"hierarchical_multitenancy": "enable",
                     "device_name": "dev2",
                     "interface_vlan_map": {
                         "device_1": {
                             "mgmt_ip_address": "192.168.90.46",
                             "ethernet_interfaces": [{
                                 "interface_num": 4,
                                 "vlan_map": [
                                     {"vlan_id": 3, "ve_ip": "192.168.190.61"},
                                     {"vlan_id": 4, "ve_ip": "172.16.1.61"},
                                     {"vlan_id": 5, "ve_ip": "192.168.91.61"},
                                 ]},
                             ]
                          }
                        }
                     },
             ]
</b>
</pre>

## Test Cases
1. Same steps as QA 
2. test vThunder loadbalancer update flow 

## Manual Testing
1. Same steps as QA 
```
vrrp-a vrid 7
  floating-ip 192.168.91.126
!
slb virtual-server 22d9c34d-844a-4ce3-92a3-49d07162f3fc 192.168.91.161
!
```

2. test vThunder loadbalancer update flow 
```
stack@openstack-4:~/source/a10-octavia/vthunder2$ openstack loadbalancer list
+--------------------------------------+------+----------------------------------+----------------+---------------------+----------+
| id                                   | name | project_id                       | vip_address    | provisioning_status | provider |
+--------------------------------------+------+----------------------------------+----------------+---------------------+----------+
| dd9c4214-5534-4444-8a28-3e6ca35be2be | vip1 | 99c9c2304f114685a32db30769c8a7e2 | 192.168.91.247 | ACTIVE              | a10      |
+--------------------------------------+------+----------------------------------+----------------+---------------------+----------+
stack@openstack-4:~/source/a10-octavia/vthunder2$ openstack loadbalancer set --description 'test' vip1
stack@openstack-4:~/source/a10-octavia/vthunder2$ openstack loadbalancer show vip1
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| created_at          | 2021-06-08T12:38:33                  |
| description         | test                                 |
| flavor_id           | None                                 |
| id                  | dd9c4214-5534-4444-8a28-3e6ca35be2be |
| listeners           |                                      |
| name                | vip1                                 |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | 99c9c2304f114685a32db30769c8a7e2     |
| provider            | a10                                  |
| provisioning_status | ACTIVE                               |
| updated_at          | 2021-06-08T13:05:16                  |
| vip_address         | 192.168.91.247                       |
| vip_network_id      | 61305531-6e3a-44b4-8e67-05e93363dddc |
| vip_port_id         | 889983ae-7da4-4419-b6bc-f290e23507d9 |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | 3ea80a1c-5785-4252-b751-c84d614ccb0c |
+---------------------+--------------------------------------+

```

### CODE REVIEW

File: a10_octavia/controller/worker/tasks/a10_network_tasks.py [hunk 1]
```diff
@@ -704,6 +704,20 @@ def _remove_device_vrid_fip(self, partition_name, vrid_value):
...
 
+    def _create_vrid(self, vrid):
+        has_vrid = True
+        try:
+            self.axapi_client.vrrpa.get(vrid)
+        except acos_errors.NotFound:
+            has_vrid = False
+
+        try:
+            if not has_vrid:
```
Comment (rahulkatre-a10) [inline a10_octavia/controller/worker/tasks/a10_network_tasks.py:715]: Why not add this block of code in above exception.

**Review by rahulkatre-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 2]
```diff
@@ -562,6 +562,63 @@ def get_update_load_balancer_flow(self, topology):
...
+        update_LB_flow.add(network_tasks.ApplyQos(
+            requires=(constants.LOADBALANCER, constants.UPDATE_DICT)))
+        update_LB_flow.add(self.handle_vrid_for_loadbalancer_subflow())
+
+        update_LB_flow.add(virtual_server_tasks.UpdateVirtualServerTask(
+            requires=(constants.LOADBALANCER, a10constants.VTHUNDER,
+                      constants.FLAVOR_DATA)))
+        update_LB_flow.add(database_tasks.UpdateLoadbalancerInDB(
+            requires=[constants.LOADBALANCER, constants.UPDATE_DICT]))
+        if CONF.a10_global.network_type == 'vlan':
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:606]: If we are adding a rack flow for update, then we can remove this vlan section from the `get_update_load_balancer_flow` used by the vthunder-amphora. The devices that live within Openstack don't need to be VLAN aware. I believe the only exception to this is when Hierarchical Portbinding is in used, but we haven't implemented it yet and it'll likely be a different flow than this one.

File: a10_octavia/controller/worker/tasks/a10_network_tasks.py (hunk 1 above)
Comment (hthompson-a10) [reply a10_octavia/controller/worker/tasks/a10_network_tasks.py:715]: A better method would be to add an `exists` function in acos-client (example: https://github.com/a10networks/acos-client/blob/6df217774ee94912a9760741c9271fa4d8abc7af/acos_client/v30/file/ssl_key.py#L30). Then we can do the following:

```
try:
    if not self.axapi_client.vrrpa.exists(vrid):
        self.axapi_client.system.action.configureVRID(vrid)
except Exception as e:
   LOG.exception("Failed to create vrid d: %s", vrid, str(e))
   raise e
```

File: a10_octavia/controller/worker/tasks/a10_network_tasks.py [hunk 3]
```diff
@@ -817,6 +831,7 @@ def execute(self, vthunder, lb_resource, vrid_list, subnet,
             if vrid_subnet.id == subnet.id or vrid.vrid_floating_ip in existing_fips:
                 vrid_floating_ips.append(vrid.vrid_floating_ip)
 
+        self._create_vrid(vrid_value)
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/tasks/a10_network_tasks.py:834]: In my opinion, we should not create the VRID here. We should do so in a different task. This way we can fail and revert if the VRID cannot be created for some reason.

We can just create a small `HandleVRID` task that runs before `HandleVRIDFloatingIP` and ensures the desired VRID exists or is created.

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/controller/worker/tasks/a10_network_tasks.py (hunk 3 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/tasks/a10_network_tasks.py]: done

**Review by rahulkatre-a10 — APPROVED** [review]

**Review by NehaKembalkarA10 — APPROVED** [review]

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py (hunk 2 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py]: done

**Pull request merged**